gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java

```diff
@@ -834,86 +834,115 @@ protected String getTopicName(TableIdentifier tid, TableMetadata tableMetadata)
+      //Update properties
+      UpdateProperties updateProperties = transaction.updateProperties();
+      props.forEach(updateProperties::set);
+      updateProperties.commit();
+      try (AutoCloseableHiveLock lock = this.locks.getTableLock(dbName, tableName);
+          Timer.Context context = new Timer().time()) {
+        transaction.commitTransaction();
+        log.info("Committing transaction for table {} took {} ms", tid, TimeUnit.NANOSECONDS.toMillis(context.stop()));
+        transactionCommitted = true;
       }
-    } catch (RuntimeException e) {
-      throw new IOException(String.format("Fail to flush table %s %s", dbName, tableName), e);
+
+      postCommit(tableMetadata, dbName, tableName, topicName, highWatermark);
+      //Reset the table metadata for next accumulation period
+      Map<String, String> currentProps = tableMetadata.table.get().properties();
+      Snapshot snapshot = tableMetadata.table.get().currentSnapshot();
+      tableMetadata.reset(currentProps, highWatermark);
+      log.info(String.format("Finish commit of new snapshot %s for table %s", snapshot.snapshotId(), tid));
     } catch (Exception e) {
-      throw new IOException(String.format("Fail to flush table %s %s", dbName, tableName), e);
+      throw new IOException(String.format("Failed to flush table %s %s. transactionCommitted=%s",
+          dbName, tableName, transactionCommitted), e);
     } finally {
       writeLock.unlock();
     }
   }
 
+  /**
+   * PostCommit operation that executes after the transaction is committed to the Iceberg table. Operations in this
+   * method are considered non-critical to the transaction and will not cause the transaction to fail if they fail,
+   * but should ideally still be executed for observability.
+   *
+   * One example of this is observability events / metrics like {@link org.apache.gobblin.metrics.GobblinTrackingEvent}.
+   * The default behavior is to emit a GTE for the commit event and a kafka audit event
+   *
+   * @param tableMetadata
+   * @param dbName
+   * @param tableName
+   * @param topicName
+   * @param highWatermark
+   * @throws IOException
+   */
+  protected void postCommit(
+      TableMetadata tableMetadata,
+      String dbName,
+      String tableName,
+      String topicName,
+      long highWatermark) throws IOException {
+    // Emit GTE for snapshot commits
+    Snapshot snapshot = tableMetadata.table.get().currentSnapshot();
+    Map<String, String> currentProps = tableMetadata.table.get().properties();
+    try (Timer.Context context = new Timer().time()) {
+      submitSnapshotCommitEvent(snapshot, tableMetadata, dbName, tableName, currentProps, highWatermark);
+      log.info("Sending snapshot commit event for {} took {} ms", topicName, TimeUnit.NANOSECONDS.toMillis(context.stop()));
+    }
+
+    try (Timer.Context context = new Timer().time()) {
+      sendAuditCounts(topicName, tableMetadata.serializedAuditCountMaps);
```

Comment:
let's try send audit count first to avoid lower counting as much as possible?

Reply:
There are 2 ways to think about it:
1. Kafka Audit is easier to check, and downstream users are more likely to use this number, so we should prioritize this step
2. Gobblin commit GTE is what we use for Gobblin monitoring and so we should prioritize our own monitoring.

Practically speaking, the most common use case is the first (as you suggest), so I think it's a good idea to emit the audit count first

Reply:
+1, and in addition to that, missing one GTE won't cause alert in our monitoring system, but missing sending audit count will cause us to have less count permanently